File: examples/temporal/callback_server.py
"""FastAPI callback receiver — converts awaithumans webhooks to Temporal signals.

This process lives between the awaithumans server (which fires the
webhook) and the Temporal workflow (which is parked waiting for a
signal). It does three things:

  1. Receive POST /awaithumans/callback with the workflow ID in the
     query string and the signed body in the request.
  2. Verify the HMAC signature (proves the webhook came from your
     awaithumans server, not an attacker).
  3. Look up the running workflow by ID and signal it.

Run with:
    uvicorn callback_server:app --host 0.0.0.0 --port 8765

For the awaithumans server (running locally) to reach this from
inside Docker / from a hosted deployment, expose it via a tunnel:
    ngrok http 8765
    export AWAITHUMANS_CALLBACK_BASE=https://<ngrok-id>.ngrok.io

The same `AWAITHUMANS_PAYLOAD_KEY` env var must be set on BOTH this
process and the awaithumans server — that's how HMAC keys derive
to the same value on both sides.
"""

from __future__ import annotations

import logging
import os
from contextlib import asynccontextmanager
from typing import AsyncIterator

from fastapi import FastAPI, HTTPException, Request
from temporalio.client import Client

from awaithumans.adapters.temporal import dispatch_signal

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("examples.temporal.callback_server")

TEMPORAL_ADDRESS = os.environ.get("TEMPORAL_ADDRESS", "localhost:7233")

# The Temporal client is a long-lived connection — opening one per
# request would burn 50ms+ on handshake. We hold a module-level
# singleton populated during the lifespan startup.
_client: Client | None = None


@asynccontextmanager
async def lifespan(_app: FastAPI) -> AsyncIterator[None]:
    global _client
    _client = await Client.connect(TEMPORAL_ADDRESS)
    logger.info("Connected to Temporal at %s", TEMPORAL_ADDRESS)
    yield
    # No explicit close — Client doesn't expose one and process exit
    # collects sockets.


app = FastAPI(lifespan=lifespan)


@app.post("/awaithumans/callback")
async def callback(request: Request, wf: str) -> dict:
    """Receive an awaithumans webhook and signal the Temporal workflow.

    `wf` is the workflow ID — the awaithumans adapter encodes it into
    `callback_url` when creating the task. We don't trust headers
    or bodies for routing; the workflow ID comes from a URL the
    workflow itself constructed."""
    if _client is None:
        raise HTTPException(status_code=503, detail="Temporal client not ready.")

    body = await request.body()
    signature = request.headers.get("x-awaithumans-signature")

    try:
        await dispatch_signal(
            temporal_client=_client,
            workflow_id=wf,
            body=body,
            signature_header=signature,
        )
    except PermissionError as exc:
        # Bad signature — security event, log + return 401.
        logger.warning("Rejected webhook with bad signature: %s", exc)
        raise HTTPException(status_code=401, detail=str(exc)) from exc
    except ValueError as exc:
        # Malformed body — return 400 so the awaithumans server's
        # logs flag the misconfig.
        logger.warning("Rejected malformed webhook: %s", exc)
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return {"ok": True}

File: examples/temporal-ts/callback-server.ts
/**
 * Callback receiver — converts awaithumans webhooks to Temporal signals.
 *
 * Mirror of `examples/temporal/callback_server.py` (Python). The two
 * receivers are interchangeable: a Python workflow can be signaled by
 * the TS receiver and vice versa, because the wire format and signal
 * naming are identical across SDKs.
 *
 * Three things this does:
 *
 *   1. Receives POST /awaithumans/callback?wf=<workflow_id>
 *   2. Verifies the HMAC signature header against PAYLOAD_KEY
 *   3. Looks up the workflow by ID and signals it with the response
 *
 * Uses Hono — small, runtime-portable web framework. Same shape would
 * work with Express, Fastify, Bun.serve, etc.
 *
 * Run with:
 *   npm run callback-server
 *
 * For the awaithumans server (running locally) to reach this from
 * Docker / a hosted deployment, expose it via a tunnel:
 *   ngrok http 8765
 *   export AWAITHUMANS_CALLBACK_BASE=https://<ngrok-id>.ngrok.io
 *
 * AWAITHUMANS_PAYLOAD_KEY must be set on BOTH this process and the
 * awaithumans server — that's how HMAC keys derive to the same value.
 */

import { serve } from "@hono/node-server";
import { Client, Connection } from "@temporalio/client";
import { dispatchSignal } from "awaithumans/temporal";
import { Hono } from "hono";

const TEMPORAL_ADDRESS = process.env.TEMPORAL_ADDRESS ?? "localhost:7233";
const PORT = Number(process.env.PORT ?? "8765");

async function main(): Promise<void> {
	const payloadKey = process.env.AWAITHUMANS_PAYLOAD_KEY;
	if (!payloadKey) {
		// In dev, `awaithumans dev` writes the key to a discovery file
		// AND exports it as an env var inside its own process. The
		// callback receiver runs separately and needs the same key —
		// either export it manually, or read the dev key file.
		console.error(
			"AWAITHUMANS_PAYLOAD_KEY is required.\n" +
				"  Dev: export AWAITHUMANS_PAYLOAD_KEY=$(cat .awaithumans/payload.key) " +
				"(from wherever you ran `awaithumans dev`).",
		);
		process.exit(1);
	}

	// Long-lived Temporal client — opening one per request would
	// burn ~50ms on handshake.
	const connection = await Connection.connect({ address: TEMPORAL_ADDRESS });
	const temporalClient = new Client({ connection });
	console.log(`[callback] connected to Temporal at ${TEMPORAL_ADDRESS}`);

	// Wrap the Temporal client in the structural interface
	// `dispatchSignal` expects — keeps the SDK adapter free of a
	// hard `@temporalio/client` import.
	const clientLike = {
		getHandle(workflowId: string) {
			return temporalClient.workflow.getHandle(workflowId);
		},
	};

	const app = new Hono();

	app.post("/awaithumans/callback", async (c) => {
		const workflowId = c.req.query("wf");
		if (!workflowId) {
			return c.json({ error: "missing wf query param" }, 400);
		}

		const body = await c.req.arrayBuffer();
		const signature = c.req.header("x-awaithumans-signature") ?? null;

		try {
			await dispatchSignal({
				temporalClient: clientLike,
				workflowId,
				body,
				signatureHeader: signature,
				payloadKey,
			});
		} catch (err) {
			const msg = (err as Error).message;
			// Bad signature is a security event; malformed body is a
			// misconfig. Both bubble to the awaithumans server's
			// webhook-failed log.
			if (msg.includes("Invalid awaithumans webhook signature")) {
				console.warn("[callback] rejected bad signature");
				return c.json({ error: msg }, 401);
			}
			console.warn("[callback] rejected:", msg);
			return c.json({ error: msg }, 400);
		}

		return c.json({ ok: true });
	});

	serve({ fetch: app.fetch, port: PORT }, (info) => {
		console.log(`[callback] listening on http://localhost:${info.port}`);
	});
}

main().catch((err) => {
	console.error("[callback] fatal:", err);
	process.exit(1);
});
